### PR TITLE
Optimistic signature verification for signed batch info

### PR DESCRIPTION
**Pull request by vusirikala**
## Description
This PR implements optimistic signature verification to reduce the time required to verify SignedBatchInfo.
When the optimistic signature verification feature flag is enabled, we will not verify these messages up front. We will accumulate the unverified messages, and when the accumulated voting power is higher than a threshold, we will aggregate all the signatures and verify the aggregated signature.
If the verification fails, we need to verify each individual signature. The ValidatorVerifier stores the list of authors that submitted bad messages, and will disable the optimistic signature verification for these malicious voters.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Breaking change
- [x] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [x] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [ ] Other (specify)

## Checklist
- [ ] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I identified and added all stakeholders and component owners affected by this change as reviewers
- [ ] I tested both happy and unhappy path of the functionality
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by danielxiangzl — APPROVED** [review]

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 1]
```diff
@@ -103,30 +126,107 @@ impl IncrementalProofState {
         Ok(())
     }
 
-    fn ready(&self, validator_verifier: &ValidatorVerifier) -> bool {
-        if self.aggregated_voting_power >= validator_verifier.quorum_voting_power() {
-            let recheck =
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/proof_coordinator.rs:108]: Earlier we were tracking `aggregated_voting_power` and using the `ready` function to check if we've reached the quorum. Now, I replaced it with `check_voting_power` that does the same thing.

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 2]
```diff
@@ -29,15 +30,17 @@ use tokio::{
 
 #[derive(Debug)]
 pub(crate) enum ProofCoordinatorCommand {
-    AppendSignature(SignedBatchInfoMsg),
+    // The bool indicates whether the signed batch info message is already verified.
+    // If the message is not verified, the coordinator is exepected to verify the message.
+    AppendSignature((SignedBatchInfoMsg, bool)),
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/proof_coordinator.rs:35]: Done.

**Review by zekun000 — COMMENTED** [review]
need rebase

**Comment by vusirikala** [comment]
> need rebase

Rebased the PR now. Ready for review.

File: consensus/src/quorum_store/proof_coordinator.rs [hunk 3]
```diff
@@ -101,28 +108,63 @@ impl IncrementalProofState {
...
+            .check_voting_power(self.signatures.keys(), check_super_majority)
+            .is_ok()
+    }
+
+    fn filter_invalid_signatures(&mut self, verifier: &ValidatorVerifier) {
+        let signatures = mem::take(&mut self.signatures);
+        self.signatures = verifier.filter_invalid_signatures(self.batch_info(), signatures);
+    }
+
+    fn try_aggregate(
```
[the remark this answers is not in the document]
Comment (vusirikala) [reply consensus/src/quorum_store/proof_coordinator.rs:126]: LedgerInfoWithUnverifiedSignatures is storing the ledger info along with its signatures.
Here, we need to store batch info along with its signatures.

[the remark this answers is not in the document]
Comment (zekun000) [reply consensus/src/quorum_store/proof_coordinator.rs:126]: that seems a perfect place to do generics. having a 
```
struct SignatureAggregator<T> {
  data: T,
  signatures: BTreeMap<AccountAddress, SignatureWithStatus>,
}

type LedgerInfoWithUnverifiedSignatures = SignatureAggregator<LedgerInfo>;
type BatchInfoWithUnverifiedSignatures = SignatureAggregator<BatchInfo>;
```

**Review by zekun000 — APPROVED** [review]
revert the config before landing

File: config/src/config/consensus_config.rs [hunk 4]
```diff
@@ -321,7 +321,7 @@ impl Default for ConsensusConfig {
             num_bounded_executor_tasks: 16,
             enable_pre_commit: true,
             max_pending_rounds_in_commit_vote_cache: 100,
-            optimistic_sig_verification: false,
+            optimistic_sig_verification: true,
```
Comment (vusirikala) [inline config/src/config/consensus_config.rs:324]:
```suggestion
            optimistic_sig_verification: false,
```

**Pull request merged**